### ISSUE TITLE
Tensor not properly saved using save_for_backward with torch.compile + custom ops

### ISSUE THREAD

**Issue by b-albar** [pytorch/pytorch#120950]
### 🐛 Describe the bug

When using a custom autograd function with torch.compile and saving tensors returned from a custom ops implementation with save_for_backward, the tensor returned appears to have changed between the forward and the backward pass. See below for some reproducing code. This not always happens, for example in the following code, it happens only when d_model=n_heads*kv_dim and break only one of the 3 tensors. Otherwise having multiple layers break all tensors (see the commented line with nn.Sequential)

```python
import torch
import torch.nn as nn
import copy

torch.library.define("knl::fwd", "(Tensor q, Tensor k, Tensor v) -> (Tensor, Tensor, Tensor)")

@torch.library.impl("knl::fwd", "default")
def cuda_knl_fwd(
    q: torch.Tensor,
    k: torch.Tensor,
    v: torch.Tensor,
):
    print(q.sum())
    print(k.sum())
    print(v.sum())

    return (q, k, v)

@torch.library.impl_abstract("knl::fwd", cuda_knl_fwd)
def meta_knl_fwd(
    q: torch.Tensor,
    k: torch.Tensor,
    v: torch.Tensor,
):

    return (torch.empty_like(q),
            torch.empty_like(k),
            torch.empty_like(v))

torch.library.define("knl::bwd", "(Tensor dout, Tensor q, Tensor k, Tensor v) -> (Tensor, Tensor, Tensor)")

@torch.library.impl("knl::bwd", "default")
def cuda_knl_bwd(
    dout: torch.Tensor,
    q: torch.Tensor,
    k: torch.Tensor,
    v: torch.Tensor,
):
    print(q.sum())
    print(k.sum())
    print(v.sum())

    return (q, k, v)

@torch.library.impl_abstract("knl::bwd", cuda_knl_bwd)
def meta_knl_bwd(
    dout: torch.Tensor,
    q: torch.Tensor,
    k: torch.Tensor,
    v: torch.Tensor,
):

    return (torch.empty_like(q),
            torch.empty_like(k),
            torch.empty_like(v))

class CallKernel(torch.autograd.Function):
    @staticmethod
    def forward(ctx, q, k, v):

        q_padded, k_padded, v_padded = torch.ops.knl.fwd(q, k, v)
        ctx.save_for_backward(q_padded, k_padded, v_padded)

        return q_padded

    @staticmethod
    def backward(ctx, dout, *args):
        q_padded, k_padded, v_padded = ctx.saved_tensors

        dq, dk, dv = torch.ops.knl.bwd(dout, q_padded, k_padded, v_padded)

        return dq, dk, dv


def call_kernel(q, k, v):
    return CallKernel.apply(q, k, v)


class Attention(nn.Module):
    def __init__(self):
        super().__init__()

        self.d_model=512
        self.n_heads=8
        self.kv_dim=64
        self.inner_dim=self.n_heads * self.kv_dim

        self.Wq = nn.Linear(self.d_model, self.inner_dim, bias=False)
        self.Wk = nn.Linear(self.d_model, self.inner_dim, bias=False)
        self.Wv = nn.Linear(self.d_model, self.inner_dim, bias=False)
        self.o = nn.Linear(self.inner_dim, self.d_model, bias=False)

    def forward(self, hidden_states):

        q = self.Wq(hidden_states)
        k = self.Wk(hidden_states)
        v = self.Wv(hidden_states)

        output = call_kernel(q, k, v)

        output = self.o(output)

        return output


model = Attention().cuda()
#model = nn.Sequential(Attention(), Attention()).cuda()
model_compiled = copy.deepcopy(model)
model_compiled = torch.compile(model_compiled)

input = torch.randn((4, 128, 512)).cuda()

print("Model fwd")
out_ref = model(input)
print("Model compiled fwd")
out = model_compiled(input)

print("Model bwd")
out_ref.sum().backward()
print("Model compiled bwd")
out.sum().backward()
```
```
Model bwd
tensor(-240.9885, device='cuda:0')
tensor(-92.2095, device='cuda:0')
tensor(205.0673, device='cuda:0')
Model compiled bwd
tensor(-240.9885, device='cuda:0')
tensor(-92.2095, device='cuda:0')
tensor(164.8126, device='cuda:0')
```

### Versions

PyTorch version: 2.3.0.dev20240229+cu121
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Arch Linux (x86_64)
GCC version: (GCC) 13.2.1 20230801
Clang version: 16.0.6
CMake version: version 3.28.3
Libc version: glibc-2.39

Python version: 3.11.7 (main, Jan 29 2024, 16:03:57) [GCC 13.2.1 20230801] (64-bit runtime)
Python platform: Linux-6.7.5-arch1-1-x86_64-with-glibc2.39
Is CUDA available: True
CUDA runtime version: 12.3.107
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA A100 80GB PCIe
GPU 1: NVIDIA A100 80GB PCIe
GPU 2: NVIDIA A100-PCIE-40GB
GPU 3: NVIDIA A100-PCIE-40GB
GPU 4: NVIDIA A100-PCIE-40GB

Nvidia driver version: 545.29.06
cuDNN version: Probably one of the following:
/usr/lib/libcudnn.so.8.9.7
/usr/lib/libcudnn_adv_infer.so.8.9.7
/usr/lib/libcudnn_adv_train.so.8.9.7
/usr/lib/libcudnn_cnn_infer.so.8.9.7
/usr/lib/libcudnn_cnn_train.so.8.9.7
/usr/lib/libcudnn_ops_infer.so.8.9.7
/usr/lib/libcudnn_ops_train.so.8.9.7
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Address sizes:                      43 bits physical, 48 bits virtual
Byte Order:                         Little Endian
CPU(s):                             128
On-line CPU(s) list:                0-127
Vendor ID:                          AuthenticAMD
Model name:                         AMD EPYC 7452 32-Core Processor
CPU family:                         23
Model:                              49
Thread(s) per core:                 2
Core(s) per socket:                 32
Socket(s):                          2
Stepping:                           0
Frequency boost:                    enabled
CPU(s) scaling MHz:                 91%
CPU max MHz:                        2350.0000
CPU min MHz:                        1500.0000
BogoMIPS:                           4702.10
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 hw_pstate ssbd mba ibrs ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 cqm rdt_a rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local clzero irperf xsaveerptr rdpru wbnoinvd amd_ppin arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif v_spec_ctrl umip rdpid overflow_recov succor smca sev sev_es
Virtualization:                     AMD-V
L1d cache:                          2 MiB (64 instances)
L1i cache:                          2 MiB (64 instances)
L2 cache:                           32 MiB (64 instances)
L3 cache:                           256 MiB (16 instances)
NUMA node(s):                       2
NUMA node0 CPU(s):                  0-31,64-95
NUMA node1 CPU(s):                  32-63,96-127
Vulnerability Gather data sampling: Not affected
Vulnerability Itlb multihit:        Not affected
Vulnerability L1tf:                 Not affected
Vulnerability Mds:                  Not affected
Vulnerability Meltdown:             Not affected
Vulnerability Mmio stale data:      Not affected
Vulnerability Retbleed:             Mitigation; untrained return thunk; SMT enabled with STIBP protection
Vulnerability Spec rstack overflow: Mitigation; Safe RET
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; Retpolines, IBPB conditional, STIBP always-on, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Not affected

Versions of relevant libraries:
[pip3] flake8==7.0.0
[pip3] gpytorch==1.11
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.26.4
[pip3] onnx==1.15.0
[pip3] pytorch-lightning==2.1.3
[pip3] pytorch-metric-learning==2.2.0
[pip3] pytorch-triton==3.0.0+901819d2b6
[pip3] torch==2.3.0.dev20240229+cu121
[pip3] torch-tb-profiler==0.4.3
[pip3] torchaudio==2.1.0+6ea1133
[pip3] torchmetrics==1.3.0
[pip3] torchvision==0.17.0
[pip3] triton==2.1.0
[conda] Could not collect

**Comment by bdhirsh** [comment]
cc @zou3519 (do we have docs for custom op authoring improvements to share?)

@b-albar looking over your implementation, one issue I see is that:

(1) the op reports as being a "functional" operator based on its schema (it returns fresh tensor outputs)
(2) However, the implementation of both the forward and the backward op returns the inputs directly. You can try "fixing" the custom op by e.g. returning `(q.clone(), k.clone(), v.clone())` in your forward impl (and giving the backward the same treatment). Altenratively, if your operator really is just a no-op that doesn't do anything and returns the inputs directly, you can represent this without a custom operator.

**Comment by b-albar** [comment]
@bdhirsh Thanks for pointing this out! I overlooked the fact that you need to return new tensors. The kernel I used was either returning a new tensor or was a no-op, depending on some condition, which explains the bug I had. Adding clone() fixed the issue.
